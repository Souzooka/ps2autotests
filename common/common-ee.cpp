#include <stdio.h>
#include <stdarg.h>
#include <string.h>
#include <kernel.h>

#define TEST_THREAD_PRIORITY   0x40
#define TEST_THREAD_STACK_SIZE 0x8000

char schedfBuffer[65536];
unsigned int schedfBufferPos = 0;

void schedf(const char *format, ...) {
	va_list args;
	va_start(args, format);
	schedfBufferPos += vsprintf(schedfBuffer + schedfBufferPos, format, args);
	va_end(args);
}

void flushschedf() {
	printf("%s", schedfBuffer);
	schedfBuffer[0] = '\0';
	schedfBufferPos = 0;
}

u32 g_testThreadDoneSema = 0;

extern int test_main(int argc, char *argv[]);

char testThreadStack[TEST_THREAD_STACK_SIZE] __attribute__ ((aligned(16)));
void testThreadProc() {
	//TODO: Print begin/end traces
	//TODO: Pass arguments from main thread to the test function
	test_main(0, NULL);
	SignalSema(g_testThreadDoneSema);
}

int main(int argc, char *argv[]) {
	//A thread is created to ensure the execution environment is the same
	//across all possible boot methods (direct hw boot, ps2link, emulator, etc.)
	
	ee_sema_t semaInfo;
	memset(&semaInfo, 0, sizeof(ee_sema_t));
	semaInfo.init_count = 0;
	semaInfo.max_count  = 1;
	semaInfo.option     = 0;
	g_testThreadDoneSema = CreateSema(&semaInfo);
	
	u32 testThreadId = 0;
	ee_thread_t threadParam;
	memset(&threadParam, 0, sizeof(ee_thread_t));
	threadParam.func               = (void*)&testThreadProc;
	threadParam.initial_priority   = TEST_THREAD_PRIORITY;
	threadParam.stack_size         = TEST_THREAD_STACK_SIZE;
	threadParam.stack              = testThreadStack;
	testThreadId = CreateThread(&threadParam);
	
	StartThread(testThreadId, NULL);
	WaitSema(g_testThreadDoneSema);
	
	return 0;
}
