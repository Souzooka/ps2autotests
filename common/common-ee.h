#ifndef PS2AUTOTESTS_COMMON_EE_H
#define PS2AUTOTESTS_COMMON_EE_H

// Defines for MSVC highlighting.  Not intended to actually compile with msc.
#ifdef _MSC_VER
#define __STDC__
#define _EE
#define __attribute__(x)
#endif

#include <stdio.h>
#include <tamtypes.h>

#undef main
#define main test_main

void schedf(const char *format, ...);
void flushschedf();

#endif
