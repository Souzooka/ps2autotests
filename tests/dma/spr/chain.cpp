#include <common-ee.h>
#include <assert.h>
#include <kernel.h>
#include <malloc.h>
#include <string.h>
#include "../dmaregs.h"
#include "../dmasend.h"
#include "../dmatags.h"

static volatile DMA::Channel *const fromSPR = DMA::D8;
static volatile DMA::Channel *const toSPR = DMA::D9;

static void clearSPR(u32 *buf) {
	memset(buf, 0, 16 * 1024);

	toSPR->sadr = 0;
	DMA::SendSimple(toSPR, buf, 16 * 1024);
}

static void testSADROffsets(u32 *buf) {
	printf("SADR offsets:\n");

	DMA::SrcChainPacket srcTag(512);

	buf[0] = 0x01234567;
	buf[1] = 0x89ABCDEF;
	buf[2] = 0xDEADBEEF;
	buf[3] = 0x1337C0DE;

	srcTag.Reset();
	srcTag.REFE(buf, 1);

	// And then set some data later into it.
	toSPR->sadr = 0x20;
	DMA::SendChain(toSPR, srcTag.Raw(), 512);

	// Now let's read it back (from channel 8.)
	fromSPR->sadr = 0x10;
	DMA::SendSimple(fromSPR, buf, 32);

	printf("  SADR updated: to=%08x, from=%08x\n", toSPR->sadr, fromSPR->sadr);
	printf("  Tag data updated: MADR=+%08x, TADR=+%08x\n", (const u8 *)toSPR->madr - (const u8 *)buf, (const u8 *)toSPR->tadr - (const u8 *)srcTag.Raw());
	printf("  SADR +0x10: %08x %08x %08x %08x\n", buf[0], buf[1], buf[2], buf[3]);
	printf("  SADR +0x20: %08x %08x %08x %08x\n", buf[4], buf[5], buf[6], buf[7]);
	printf("\n");

	// How about misaligned SADR values?
	memset(buf, 0xCC, 16 * 1024);
	buf[0] = 0xDEADBEEF;
	buf[1] = 0x89ABCDEF;
	buf[2] = 0x01234567;
	buf[3] = 0x1337C0DE;

	srcTag.Reset();
	srcTag.REFE(buf, 1);

	toSPR->sadr = 0x14;
	DMA::SendChain(toSPR, srcTag.Raw(), 512);

	memset(buf, 0xCC, 16 * 1024);

	fromSPR->sadr = 0x10;
	DMA::SendSimple(fromSPR, buf, 32);

	printf("  SADR updated: to=%08x, from=%08x\n", toSPR->sadr, fromSPR->sadr);
	printf("  Tag data updated: MADR=+%08x, TADR=+%08x\n", (const u8 *)toSPR->madr - (const u8 *)buf, (const u8 *)toSPR->tadr - (const u8 *)srcTag.Raw());
	printf("  SADR +0x14: %08x %08x %08x %08x\n", buf[0], buf[1], buf[2], buf[3]);
	printf("  SADR +0x24: %08x %08x %08x %08x\n", buf[4], buf[5], buf[6], buf[7]);

	fromSPR->sadr = 0;
	toSPR->sadr = 0;
}

static void testSADRHighBits(u32 *buf) {
	printf("SADR high bits:\n");

	DMA::SrcChainPacket srcTag(512);

	buf[0] = 0x01234567;
	buf[1] = 0x89ABCDEF;
	buf[2] = 0xDEADBEEF;
	buf[3] = 0x1337C0DE;

	srcTag.Reset();
	srcTag.REFE(buf, 16);

	toSPR->sadr = 0x81234028;
	DMA::SendChain(toSPR, srcTag.Raw(), 512);

	memset(buf, 0, 16);

	fromSPR->sadr = 0x00000020;
	DMA::SendSimple(fromSPR, buf, 16);

	printf("  SADR updated: to=%08x, from=%08x\n", toSPR->sadr, fromSPR->sadr);
	printf("  Tag data updated: MADR=+%08x, TADR=+%08x\n", (const u8 *)toSPR->madr - (const u8 *)buf, (const u8 *)toSPR->tadr - (const u8 *)srcTag.Raw());
	printf("  SADR send with ignored bits: %08x %08x %08x %08x\n", buf[0], buf[1], buf[2], buf[3]);
	printf("\n");

	memset(buf, 0, 16);

	fromSPR->sadr = 0x6703802C;
	DMA::SendSimple(fromSPR, buf, 16);

	printf("  SADR updated: to=%08x, from=%08x\n", toSPR->sadr, fromSPR->sadr);
	printf("  SADR receive with ignored bits: %08x %08x %08x %08x\n", buf[0], buf[1], buf[2], buf[3]);

	fromSPR->sadr = 0;
	toSPR->sadr = 0;
}

static void testSizeZero(u32 *buf) {
	printf("Size zero:\n");

	DMA::SrcChainPacket srcTag(512);

	memset(buf, 0xDD, 16 * 1024);

	buf[0] = 0x01234567;
	buf[1] = 0x89ABCDEF;
	buf[2] = 0xDEADBEEF;
	buf[3] = 0x1337C0DE;

	srcTag.Reset();
	srcTag.REFE(buf, 0);

	toSPR->sadr = 0;
	DMA::SendChain(toSPR, srcTag.Raw(), 512);

	SyncDCache(buf, (u8 *)buf + 16 * 1024);
	memset(buf, 0xCC, 16 * 1024);
	SyncDCache(buf, (u8 *)buf + 16 * 1024);

	fromSPR->sadr = 0;
	DMA::SendSimple(fromSPR, buf, 16 * 1024);

	SyncDCache(buf, (u8 *)buf + 16 * 1024);

	printf("  SADR updated: to=%08x, from=%08x\n", toSPR->sadr, fromSPR->sadr);
	printf("  Send zero 0x0000: %08x %08x %08x %08x\n", buf[0], buf[1], buf[2], buf[3]);
	printf("  Send zero 0x0010: %08x %08x %08x %08x\n", buf[4], buf[5], buf[6], buf[7]);

	for (u32 i = 16 / sizeof(u32); i < 16 * 1024 / sizeof(u32); ++i) {
		if (buf[i] != 0) {
			printf("  Read zeros until: 0x%04x %08x\n", i * sizeof(u32), buf[i]);
			break;
		}
	}

	fromSPR->sadr = 0;
	toSPR->sadr = 0;
}

int main(int argc, char *argv[]) {
	printf("-- TEST BEGIN\n");

	// We'll reuse this buffer for the SPR in all the tests.
	u32 *buf = (u32 *)memalign(16, 16 * 1024);

	// Offsets should work.
	clearSPR(buf);
	testSADROffsets(buf);
	printf("\n");

	// Certain bits of SADR are ignored.
	clearSPR(buf);
	testSADRHighBits(buf);
	printf("\n");

	// Setting the SPR bit in chains for fromSPR/toSPR seems to cause crashes.

	// How many bytes does a zero-sized transfer copy?
	clearSPR(buf);
	testSizeZero(buf);

	free(buf);

	printf("-- TEST END\n");
	return 0;
}
