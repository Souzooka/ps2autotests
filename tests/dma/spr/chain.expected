-- TEST BEGIN
SADR offsets:
  SADR updated: to=00000030, from=00000030
  Tag data updated: MADR=+00000010, TADR=+00000010
  SADR +0x10: 00000000 00000000 00000000 00000000
  SADR +0x20: 01234567 89abcdef deadbeef 1337c0de

  SADR updated: to=00000020, from=00000030
  Tag data updated: MADR=+00000010, TADR=+00000010
  SADR +0x14: deadbeef 89abcdef 01234567 1337c0de
  SADR +0x24: 01234567 89abcdef deadbeef 1337c0de

SADR high bits:
  SADR updated: to=00000120, from=00000030
  Tag data updated: MADR=+00000100, TADR=+00000010
  SADR send with ignored bits: 01234567 89abcdef deadbeef 1337c0de

  SADR updated: to=00000120, from=00000030
  SADR receive with ignored bits: 01234567 89abcdef deadbeef 1337c0de

Size zero:
  SADR updated: to=00000000, from=00000000
  Send zero 0x0000: 00000000 00000000 00000000 00000000
  Send zero 0x0010: 00000000 00000000 00000000 00000000
-- TEST END
